test/libsolidity/SolidityEndToEndTest.cpp

```diff
@@ -10589,6 +10589,535 @@ BOOST_AUTO_TEST_CASE(snark)
+{
+	char const* sourceCode = R"(
+	contract test {
+		function f(int8 x) public returns (fixed256x19) {
+			return x;
+		}
+	}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	testContractAgainstCppOnRange("f(int8)", [](u256 const &a) -> u256 { return a * pow(u256(10), 19); }, -50, -1);
+	testContractAgainstCppOnRange("f(int8)", [](u256 const &a) -> u256 { return a * pow(u256(10), 19); }, 0, 50);
+}
+
+BOOST_AUTO_TEST_CASE(ufixed_uint_conversion)
+{
+	char const* sourceCode = R"(
+	contract test {
+		function f(ufixed256x19 x) public returns (uint256) {
+			return uint256(x);
+		}
+	}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	testContractAgainstCppOnRange("f(ufixed256x19)", [](u256 const &a) -> u256 { return 0 + 0 * a; }, pow(u256(10), 19) - 10, pow(u256(10), 19));
+	testContractAgainstCppOnRange("f(ufixed256x19)", [](u256 const &a) -> u256 { return 1 + 0 * a; }, pow(u256(10), 19), pow(u256(10), 19) + 10);
+	testContractAgainstCppOnRange("f(ufixed256x19)", [](u256 const &a) -> u256 { return 2 + 0 * a; }, 2 * pow(u256(10), 19), 2 * pow(u256(10), 19) + 10);
+}
+
+BOOST_AUTO_TEST_CASE(fixed_int_conversion)
+{
+	char const* sourceCode = R"(
+	contract test {
+		function f(fixed256x19 x) public returns (int256) {
+			return int256(x);
+		}
+	}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	testContractAgainstCppOnRange("f(fixed256x19)", [](u256 const &a) -> u256 { return createFixed(0) + 0 * a; }, createFixed(-1, 1, 19) + 1, createFixed(-1, 1, 19) + 10);
+	testContractAgainstCppOnRange("f(fixed256x19)", [](u256 const &a) -> u256 { return createFixed(-1) + 0 * a; }, createFixed(-1, 1, 19) - 10, createFixed(-1, 1, 19));
+	testContractAgainstCppOnRange("f(fixed256x19)", [](u256 const &a) -> u256 { return createFixed(-2) + 0 * a; }, createFixed(-2, 1, 19) - 10, createFixed(-2, 1, 19));
+
+	testContractAgainstCppOnRange("f(fixed256x19)", [](u256 const &a) -> u256 { return 0 + 0 * a; }, pow(u256(10), 19) - 10, pow(u256(10), 19));
+	testContractAgainstCppOnRange("f(fixed256x19)", [](u256 const &a) -> u256 { return 1 + 0 * a; }, pow(u256(10), 19), pow(u256(10), 19) + 10);
+	testContractAgainstCppOnRange("f(fixed256x19)", [](u256 const &a) -> u256 { return 2 + 0 * a; }, 2 * pow(u256(10), 19), 2 * pow(u256(10), 19) + 10);
+}
+
+BOOST_AUTO_TEST_CASE(ufixed_overflow)
+{
+	char const* sourceCode = R"(
+	contract test {
+		ufixed256x19 a = (2**256-1)/(10**19);
+		function f() returns (ufixed256x19) {
+			return a + 1e-19;
+		}
+	}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	ABI_CHECK(callContractFunction("f()"), encodeArgs(u256(0)));
+}
+
+BOOST_AUTO_TEST_CASE(fixed_overflow)
+{
+	char const* sourceCode = R"(
+	contract test {
+		fixed256x19 a = fixed256x19(ufixed256x19((2**255-1)/(10**19)));
+		function f() returns (fixed256x19) {
+			return a + 1e-19;
+		}
+	}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	ABI_CHECK(callContractFunction("f()"), encodeArgs(pow(u256(2), 255)));
+}
+
+BOOST_AUTO_TEST_CASE(short_ufixed_overflow)
```

Comment:
Can you also add a test for multiplication overflow and perhaps also division?

I guess we even support `%` which we either shouldn't or should be clear on its semantics.

Reply:
I'll add those tests.

Modulo is well defined on non-integer numbers. It's defined in the same way as integer modulo. Equivalently, you just multiply both operands up so that they are integers, then divide the result by whatever you multiplied by in the beginning. There is done by the fixed point factor.

Reply:
ok, thanks! Then please add a test :)